### DISCUSSION TITLE
QC for Protocol Deviation Assessment (PD)

### DISCUSSION THREAD

**Discussion by mattroumaya** [Gilead-Public/gsm#209] [General]
# QC for Protocol Deviation Assessment (PD)

- [ ] Documentation
    - [ ] Function name captured in [Roxygen title](https://cran.r-project.org/web/packages/roxygen2/vignettes/rd.html#the-description-block)(e.g. "Adverse Event Assessment")
    - [ ] Assessment Purpose captured in [Roxygen description](https://cran.r-project.org/web/packages/roxygen2/vignettes/rd.html#the-description-block) (e.g. "Evaluates site-level level Adverse Event Rates and flags rates that are abnormally high or low."
    - [ ] Input Data Requirements are described in the the [Roxygen details](https://cran.r-project.org/web/packages/roxygen2/vignettes/rd.html#the-description-block) under a Data section (`#' @details # Data Requirements`).
    - [x] Statistical Assumptions - (Assess() and Analyze() only) List the statistical methods used and refer to detailed description in Analyze function in a the [Roxygen details](https://cran.r-project.org/web/packages/roxygen2/vignettes/rd.html#the-description-block) under a Statistics section (`#' @details # Statistical Assumptions`).
    - [ ] Parameters have detailed documentation using `@param`. Each parameter should include: parameter name, brief description, usage details, the default value (if any), is it required, list of valid options (if applicable)
    - [x] All external dependencies are captured. Use `@importFrom` when importing 3 or fewer functions, and use `@import` otherwise. 
    - [x] Output Data Standards are described under `@returns`
    - [x] At least 1 example is provided under `@examples`
- [ ] Error Checking
    - [ ]  Basic checks for all all parameters should be included using `stopifnot()` or similar logic. (e.g. `stopifnot("dfInput is not a dataframe"=is.data.frame(dfInput))`)
    - [ ] Tests confirm that `stopifnot()` parameter checks are working as expected. 
    - [ ] Tests confirm that the input data has required columns (if any)
    - [x] Tests confirm that the output data has the expected columns
    - [ ] Tests confirm intended functionality for each parameter
    - [ ] Tests confirm that missing data in required columns is handled appropriately and errors/warnings are produced if needed. 
- [ ] Basic QC
    - [ ] Assessment has User Requirements + Qualification tests captured using {valtools} framework. Report is generating as expected and all checks are passing. 
    - [x] Code is well commented and easy to read
    - [x] No file paths or other company-specific data is present
    - [x] Function calls from non-tidyverse dependencies are called via `::`
    - [x] devtools::check() passes with no errors/warnings/notes
    - [x] package docs up to date. devtools::document() doesn't change any files
    - [x] Uses tidyverse best practices for standard data manipulation (start discussion thread if unclear)
    - [x] All new dependencies add significant value (start discussion thread)
    - [x] All GitHub Actions run with no errors

**Comment by mattroumaya** [comment]
## PD_Map_Raw()

**Parameters have detailed documentation using @param. Each parameter should include: parameter name, brief description, usage details, the default value (if any), is it required, list of valid options (if applicable)**
- Specify _PD dataset with required SUBJID column_ [here](https://github.com/Gilead-BioStats/gsm/blob/dev/R/PD_Map_Raw.R#L23)

**Basic checks for all all parameters should be included using stopifnot() or similar logic. (e.g. stopifnot("dfInput is not a dataframe"=is.data.frame(dfInput)))**
- Minor suggestion: change `Time on Study dataset` --> `RDSL dataset` [here](https://github.com/Gilead-BioStats/gsm/blob/dev/R/PD_Map_Raw.R#L40)

**Tests confirm that stopifnot() parameter checks are working as expected.**
- Need test for `strExposureCol`
  - Suggest: `expect_error()` test when providing wrong type

```r
expect_error(
PD_Map_Raw(clindata::raw_protdev,clindata::rawplus_rdsl, strExposureCol = 123)
)
```
**Tests confirm intended functionality for each parameter**
- Same as above - need test for `strExposureCol`

---

## PD_Assess()

**Function name captured in Roxygen title(e.g. "Adverse Event Assessment")**
- [Definition](https://github.com/Gilead-BioStats/gsm/blob/dev/R/PD_Assess.R#L1) is _Protocol Deviation Assessment using Poisson Regression_
  - Should this change to either: 1) Protocol Deviation Assessment using Poisson or Wilcoxon Regression or; 2) Protocol Deviation Assessment
  - Or leave as is?

**Assessment Purpose captured in Roxygen description (e.g. "Evaluates site-level level Adverse Event Rates and flags rates that are abnormally high or low."**
- Contains **TODO**: _The Protocol Deviation Assessment uses the standard GSM data pipeline (**TODO add link to data vignette**) to flag possible outliers. More details regarding the data pipeline and statistical methods are described below._ 

**Input Data Requirements are described in the the Roxygen details under a Data section (#' @details # Data Requirements).**
- Missing description for `Count` column [here](https://github.com/Gilead-BioStats/gsm/blob/dev/R/PD_Assess.R#L9-L14)

**Statistical Assumptions - (Assess() and Analyze() only) List the statistical methods used and refer to detailed description in Analyze function in a the Roxygen details under a Statistics section (#' @details # Statistical Assumptions).**
- Minor fix: "their" is misspelled [here](https://github.com/Gilead-BioStats/gsm/blob/dev/R/PD_Assess.R#L25) at the end of the line

**Parameters have detailed documentation using @param. Each parameter should include: parameter name, brief description, usage details, the default value (if any), is it required, list of valid options (if applicable)**
- [`dfInput`](https://github.com/Gilead-BioStats/gsm/blob/dev/R/PD_Assess.R#L27) needs to specify that `Rate` column is required.
  - Suggested fix: _dfInput input data with one record per person and the following required columns: SubjectID, SiteID, Count, Exposure, Rate_
- [`vThreshold`](https://github.com/Gilead-BioStats/gsm/blob/dev/R/PD_Assess.R#L28) - add quotes around `"wilcoxon"` and make W lower-case
- [`bDataList`](https://github.com/Gilead-BioStats/gsm/blob/dev/R/PD_Assess.R#L32) - suggested change to: _If FALSE (the default), only the summary data frame is returned_

**Basic checks for all all parameters should be included using stopifnot() or similar logic. (e.g. stopifnot("dfInput is not a dataframe"=is.data.frame(dfInput)))**
- Need check for [`nCutoff`](https://github.com/Gilead-BioStats/gsm/blob/dev/R/PD_Assess.R#L29)
  - `is.numeric(nCutoff)`
  - `length(nCutoff) == 1`

- Need check for [`cLabel`](https://github.com/Gilead-BioStats/gsm/blob/dev/R/PD_Assess.R#L30)
  - `length(cLabel) == 1`

- Need check for [`cMethod`](https://github.com/Gilead-BioStats/gsm/blob/dev/R/PD_Assess.R#L31)
  - `length(cMethod) == 1`

- Need checks for required input columns
- Need checks for NA handling and potentially duplicate ID handling

**Tests confirm that stopifnot() parameter checks are working as expected.**
- Need tests for `vThreshold` and `nCutoff`

**Tests confirm that the input data has required columns (if any)**
- No issues, but can combine [lines 16 - 29](https://github.com/Gilead-BioStats/gsm/blob/dev/tests/testthat/test_PD_Assess.R#L16-L29)

**Tests confirm intended functionality for each parameter**
- Need tests for `nCutoff`, `cLabel`, `cMethod`, `NA` handling, as noted above

@jwildfire ready for your review

**Reply by jwildfire** [reply]
> ## PD_Map_Raw()
> **Parameters have detailed documentation using @param. Each parameter should include: parameter name, brief description, usage details, the default value (if any), is it required, list of valid options (if applicable)**
>
> * Specify _PD dataset with required SUBJID column_ [here](https://github.com/Gilead-BioStats/gsm/blob/dev/R/PD_Map_Raw.R#L23)
>
> **Basic checks for all all parameters should be included using stopifnot() or similar logic. (e.g. stopifnot("dfInput is not a dataframe"=is.data.frame(dfInput)))**
> 
> * Minor suggestion: change `Time on Study dataset` --> `RDSL dataset` [here](https://github.com/Gilead-BioStats/gsm/blob/dev/R/PD_Map_Raw.R#L40)
> 
> **Tests confirm that stopifnot() parameter checks are working as expected.**
> 
> * Need test for `strExposureCol`
>   
>   * Suggest: `expect_error()` test when providing wrong type
> 
> ```r
> expect_error(
> PD_Map_Raw(clindata::raw_protdev,clindata::rawplus_rdsl, strExposureCol = 123)
> )
> ```
> 
> **Tests confirm intended functionality for each parameter**
> 
> * Same as above - need test for `strExposureCol`

👍  these all make sense to me.

**Reply by jwildfire** [reply]
> ## PD_Assess()
> **Function name captured in Roxygen title(e.g. "Adverse Event Assessment")**
> 
> * [Definition](https://github.com/Gilead-BioStats/gsm/blob/dev/R/PD_Assess.R#L1) is _Protocol Deviation Assessment using Poisson Regression_
>   
>   * Should this change to either: 1) Protocol Deviation Assessment using Poisson or Wilcoxon Regression or; 2) Protocol Deviation Assessment
>   * Or leave as is?

👍  I vote for option 2 above. 


> **Assessment Purpose captured in Roxygen description (e.g. "Evaluates site-level level Adverse Event Rates and flags rates that are abnormally high or low."**
> 
> * Contains **TODO**: _The Protocol Deviation Assessment uses the standard GSM data pipeline (**TODO add link to data vignette**) to flag possible outliers. More details regarding the data pipeline and statistical methods are described below._

Ok for now. Will fix with #83 

> **Input Data Requirements are described in the the Roxygen details under a Data section (#' @details # Data Requirements).**
> 
> * Missing description for `Count` column [here](https://github.com/Gilead-BioStats/gsm/blob/dev/R/PD_Assess.R#L9-L14)

👍  typo needs fix. 

> **Statistical Assumptions - (Assess() and Analyze() only) List the statistical methods used and refer to detailed description in Analyze function in a the Roxygen details under a Statistics section (#' @details # Statistical Assumptions).**
> 
> * Minor fix: "their" is misspelled [here](https://github.com/Gilead-BioStats/gsm/blob/dev/R/PD_Assess.R#L25) at the end of the line

👍  If i head a dollar for every time i've spelled that wrong ... 

> **Parameters have detailed documentation using @param. Each parameter should include: parameter name, brief description, usage details, the default value (if any), is it required, list of valid options (if applicable)**
> 
> * [`dfInput`](https://github.com/Gilead-BioStats/gsm/blob/dev/R/PD_Assess.R#L27) needs to specify that `Rate` column is required.
>   
>   * Suggested fix: _dfInput input data with one record per person and the following required columns: SubjectID, SiteID, Count, Exposure, Rate_
> * [`vThreshold`](https://github.com/Gilead-BioStats/gsm/blob/dev/R/PD_Assess.R#L28) - add quotes around `"wilcoxon"` and make W lower-case
> * [`bDataList`](https://github.com/Gilead-BioStats/gsm/blob/dev/R/PD_Assess.R#L32) - suggested change to: _If FALSE (the default), only the summary data frame is returned_

👍 
 
> **Basic checks for all all parameters should be included using stopifnot() or similar logic. (e.g. stopifnot("dfInput is not a dataframe"=is.data.frame(dfInput)))**
> 
> * Need check for [`nCutoff`](https://github.com/Gilead-BioStats/gsm/blob/dev/R/PD_Assess.R#L29)
>   
>   * `is.numeric(nCutoff)`
>   * `length(nCutoff) == 1`
> * Need check for [`cLabel`](https://github.com/Gilead-BioStats/gsm/blob/dev/R/PD_Assess.R#L30)
>   
>   * `length(cLabel) == 1`
> * Need check for [`cMethod`](https://github.com/Gilead-BioStats/gsm/blob/dev/R/PD_Assess.R#L31)
>   
>   * `length(cMethod) == 1`
> * Need checks for required input columns
> * Need checks for NA handling and potentially duplicate ID handling

👍  Skip NA and duplicate work for now, but everything else looks good. 

> **Tests confirm that stopifnot() parameter checks are working as expected.**
> 
> * Need tests for `vThreshold` and `nCutoff`

👍 

> **Tests confirm that the input data has required columns (if any)**
> 
> * No issues, but can combine [lines 16 - 29](https://github.com/Gilead-BioStats/gsm/blob/dev/tests/testthat/test_PD_Assess.R#L16-L29)

👍 

> **Tests confirm intended functionality for each parameter**
> 
> * Need tests for `nCutoff`, `cLabel`, `cMethod`, `NA` handling, as noted above

👍

**Comment by jwildfire** [comment]
All looks good. Go ahead and file issues for v0.2
